playground-common/src/Schema.hs

```diff
@@ -0,0 +1,218 @@
+{-# LANGUAGE AllowAmbiguousTypes        #-}
+{-# LANGUAGE DefaultSignatures          #-}
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RecordWildCards            #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE TypeOperators              #-}
+{-# LANGUAGE TypeSynonymInstances       #-}
+
+module Schema
+    ( ToSchema
+    , toSchema
+    , TypeSignature(..)
+    , toTypeSignature
+    , typeSignature
+    , DataType(..)
+    , Constructor(..)
+    , ConstructorName(..)
+    ) where
+
+import           Control.Newtype.Generics (Newtype)
+import           Data.Aeson               (FromJSON, ToJSON)
+import           Data.Map                 (Map)
+import           Data.Monoid              ((<>))
+import           Data.Proxy               (Proxy (Proxy))
+import           Data.String              (IsString)
+import           Data.Text                (Text)
+import qualified Data.Text                as Text
+import           Data.Typeable            (splitTyConApp)
+import           GHC.Generics             ((:*:) ((:*:)), (:+:), C1, D1, Generic, M1 (M1), Rec0, Rep, S1, Selector, U1,
+                                           V1, conIsRecord, conName, from, selName)
+import qualified GHC.Generics             as GHC
+import           Type.Reflection          (SomeTypeRep, Typeable, someTypeRep, tyConModule, tyConName)
+
+{-# ANN module ("HLint: ignore Avoid restricted function" :: Text)
+        #-}
+
+data TypeSignature =
+    TypeSignature
+        { moduleName         :: Text
+        , constructorName    :: Text
+        , argumentSignatures :: [TypeSignature]
+        }
+    deriving (Eq, Show, Ord, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+data DataType =
+    DataType TypeSignature [Constructor]
+    deriving (Eq, Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+data Constructor
+    = Constructor ConstructorName [DataType]
```

Comment:
This seems like we have awkward mutual recursion between `DataType` and `Constructor`, do we not want some kind of type reference thing?